diesel/src/pg/types/date_and_time/time.rs

```diff
@@ -44,6 +49,23 @@ impl ToSql<Timestamp, Pg> for PrimitiveDateTime {
     }
 }
 
+// Delegate offset datetimes in terms of UTC primitive datetimes; this stores everything in the DB as UTC
+#[cfg(all(feature = "time", feature = "postgres_backend"))]
+impl ToSql<Timestamp, Pg> for OffsetDateTime {
```

Comment:
While strictly true, this winds up being incredibly verbose in practice. In particular where users of `diesel` don't control their own schema, it's convenient to specify the UTC convention at the row mapping. This is probably why the MySql impl exists.

When deriving `Queryable` for large structs on `Pg` you currently have to write the field mapping as `PrimitiveDateTime`. When doing almost anything with the data later (including serializing for REST output) you wind up with a lot of `assume_utc()` calls. This becomes confusing to read, because "hey this Postgres time is in UTC" annotation is separated from the database call that populated the `PrimitiveDateTime`.

In the other direction when storing rows, you wind up having to write wrapper fns to manually convert `OffsetDateTime` into `PrimitiveDateTime` for each and every insertion. This is easy to get wrong: if you simply call `date()` and `time()` without first explicitly converting to UTC you can easily wind up inadvertently storing an incorrect timestamp. There is no impl in `time` to perform this conversion, making it almost deliberately awkward. I assume the `time` crate author has a philosophical reasoning here, in that most people should store and operate on `OffsetDateTime` instead of `PrimitiveDateTime`.

Reply:
In this case where you don't control the actual database schema, I think you have two reasonable options:
- Use `serialize_as`/[`serialize_fn`](https://github.com/diesel-rs/diesel/pull/3837)/`deserialize_as`/`deserialize_fn` on your Queryable struct - this removes the need for a conversion at every struct construction (resp. usage)
- Since it turns out that the actual wire serialization for the `Timestamptz` type is precisely that of serializing a `Timestamp` as `Utc`, you could update your `schema.rs` to replace `Timestamp` by `Timestamptz` there after the CLI generates it, in effect asserting that when a `Timestamp` is used in this table, it is indeed expected to be an `UTC` timestamp. You would then appropriately use timezone-aware throughout your rust application, guaranteeing that these are read and inserted as such, and avoiding non-timezone-aware types in your application (which is best in any case). (This is a bit of a hack but I think it's a good compromise)

Reply:
This is interesting. I didn't realize the Timestamp behavior was that different, but I guess it does make sense here.

> you could update your schema.rs to replace Timestamp by Timestamptz there after the CLI generates it, in effect asserting that when a Timestamp is used in this table, it is indeed expected to be an UTC timestamp.

This is a really clever approach. I was able to make this switch in my own code, and things generally worked as you've described. The only place I ran into issues was with a `gt` comparison against `now`... to me that indicates the `NOW()` typing in Postgres actually isn't correct either. [The Postgres documentation](https://www.postgresql.org/docs/current/functions-datetime.html) seems to indicate that the type returned should be `Timestamptz` and not naked `Timestamp`. I was able to work around it by doing a local `now_utc` and doing a local subtraction, but probably something to add to a corrections list.

Reply:
IIRC `now` has `AsExpression` implementations for both `Timestamp` and `Timestamptz`, so one shouldn't hit any error if `.into_sql()`ing it to `Timestamptz`. I'm surprised that doesn't work implicitly anymore... It's maybe a matter of Nullable/not Nullable...